File: BarrelJump JavaScript Game/Obstacle.js
class Obstacle {
    constructor(parentElement, x, y, width, height, imgSrc) {
        this.parentElement = parentElement;
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
        this.imgSrc = imgSrc;

        this.imgElement = new Image();
        this.imgElement.src = this.imgSrc;
        this.imgElement.style.position = 'absolute';
        this.imgElement.style.left = `${this.x}px`;
        this.imgElement.style.top = `${this.y}px`;
        this.imgElement.style.width = `${this.width}px`;
        this.imgElement.style.height = `${this.height}px`;

        this.parentElement.appendChild(this.imgElement);
    }

    getBoundingBox() {
        return {
            left: this.x + this.width - 65,
            top: this.y + this.height - 80,
            right: this.x + 65,
            bottom: this.y + 80
        };
    }
}

export { Obstacle };

File: Kotlin Projects/JumpandRun Simple Kotlin Game/app/src/main/java/com/example/jumpandrun/Obstacle.kt
package com.example.jumpandrun

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Color

open class Obstacle(
    val x: Float,
    var y: Float,
    val width: Float,
    val height: Float
) {
    protected val paint: Paint = Paint().apply {
        color = Color.YELLOW
        style = Paint.Style.FILL
    }

    // For collision smoothing
    // private val collisionTolerance = 5f // This might not be needed or could be handled differently in subclasses

    open fun draw(canvas: Canvas) {
        canvas.drawRect(x, y, x + width, y + height, paint)
    }

    open fun handleCollision(player: Player) {
        // Calculate player boundaries
        val playerLeft = player.x
        val playerRight = player.x + player.size
        val playerTop = player.y
        val playerBottom = player.y + player.size

        // Calculate obstacle boundaries
        val obstacleLeft = x
        val obstacleRight = x + width
        val obstacleTop = y
        val obstacleBottom = y + height

        // Check if player is colliding with obstacle
        if (playerRight > obstacleLeft &&
            playerLeft < obstacleRight &&
            playerBottom > obstacleTop &&
            playerTop < obstacleBottom) {

            // Calculate overlap on all sides
            val overlapLeft = playerRight - obstacleLeft
            val overlapRight = obstacleRight - playerLeft
            val overlapTop = playerBottom - obstacleTop
            val overlapBottom = obstacleBottom - playerTop

            // Find the smallest overlap
            val minOverlap = minOf(overlapLeft, overlapRight, overlapTop, overlapBottom)

            // Resolve collision based on smallest overlap
            when (minOverlap) {
                overlapTop -> {
                    // Player hits top of obstacle (standing on it)
                    player.y = obstacleTop - player.size

                    // Only stop vertical movement if player is falling down
                    if (player.velocityY > 0) {
                        player.velocityY = 0f
                    }

                    player.onGround = true
                    player.isOnPlatform = true
                    player.collidingBottom = true

                    player.landOnPlatform()
                }
                overlapBottom -> {
                    // Player hits bottom of obstacle (head bump)
                    player.y = obstacleBottom

                    // Only stop vertical movement if player is moving up
                    if (player.velocityY < 0) {
                        player.velocityY = 0f
                    }

                    player.collidingTop = true
                }
                overlapLeft -> {
                    // Player hits left side of obstacle
                    player.x = obstacleLeft - player.size

                    // Only stop horizontal movement if player is moving right
                    if (player.velocityX > 0) {
                        player.velocityX = 0f
                    }

                    player.collidingRight = true
                }
                overlapRight -> {
                    // Player hits right side of obstacle
                    player.x = obstacleRight

                    // Only stop horizontal movement if player is moving left
                    if (player.velocityX < 0) {
                        player.velocityX = 0f
                    }

                    player.collidingLeft = true
                }
            }
        }
    }
}
